File: playground/algorithms/min-max/c/src/min_max.c
#include "min_max.h"

bool find_min_max_builtin(const int *values, size_t count, int *minimum,
                          int *maximum) {
  if (count == 0) {
    return false;
  }

  *minimum = values[0];
  *maximum = values[0];

  for (size_t i = 1; i < count; ++i) {
    if (values[i] < *minimum) {
      *minimum = values[i];
    } else if (values[i] > *maximum) {
      *maximum = values[i];
    }
  }

  return true;
}

bool find_min_max_manual(const int *values, size_t count, int *minimum,
                         int *maximum) {
  if (count == 0) {
    return false;
  }

  *minimum = values[0];
  *maximum = values[0];

  for (size_t i = 1; i < count; ++i) {
    if (values[i] < *minimum) {
      *minimum = values[i];
    } else if (values[i] > *maximum) {
      *maximum = values[i];
    }
  }

  return true;
}

File: playground/algorithms/min-max/rust/tests/min_max.rs
use min_max::{find_min_max_builtin, find_min_max_manual};

#[test]
fn find_min_max_builtin_returns_correct_min_and_max() {
    assert_eq!(find_min_max_builtin(&[1, 2, 3, 4, 5]), Some((1, 5)));
}

#[test]
fn find_min_max_manual_returns_correct_min_and_max() {
    assert_eq!(find_min_max_manual(&[1, 2, 3, 4, 5]), Some((1, 5)));
}
